code/android-core-library/src/main/java/com/adobe/marketing/mobile/utils/RemoteDownloader.kt

```diff
@@ -0,0 +1,116 @@
+/*
+  Copyright 2022 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+ */
+
+package com.adobe.marketing.mobile.utils
+
+import com.adobe.marketing.mobile.LoggingMode
+import com.adobe.marketing.mobile.MobileCore
+import com.adobe.marketing.mobile.internal.utility.StringUtils
+import com.adobe.marketing.mobile.services.CacheFileService
+import com.adobe.marketing.mobile.services.Networking
+import java.io.File
+
+/**
+ * Facilitates downloading content from a URL into a local cache directory.
+ * Stores downloaded content in to a directory governed by [CacheFileService].
+ *
+ * Currently, triggers to [download] are independent, successive calls.
+ * This class is intentionally final and components are required to use this via composition;
+ * This allows easier path for components using this class to add queuing logic for jobs internally,
+ * maintaining compatibility when this class adds queuing logic in the future, and facilitating
+ * cleaner tests in composing components.
+ */
+class RemoteDownloader(
+    private val networkService: Networking,
+    private val cacheFileService: CacheFileService
+) {
+
+    companion object {
+        private const val TAG = "RemoteDownloader"
+    }
+
+    /**
+     * Represents a component that provides metadata about file content.
+     */
+    interface MetadataProvider {
+        companion object MetadataKeys {
```

Comment:
[discussion] I think this is the right way to hold constants in Kotlin. At the same time, the public `RemoteDownloader` class and this public `MetadataProvider` interface may be used in other extensions' java code, so it may be worth time to discuss if we only expose public utility classes/interfaces as Java class or Java interface.
cc @praveek